proofs/alf/cvc5/programs/Strings.smt3

```diff
@@ -587,3 +587,147 @@
+          (alf.not isUnder)
+        (alf.ite (alf.is_eq n (- ls lt))
+          isUnder
+          false)))))))
+    (($str_arith_entail_is_approx_len (str.from_int n1) n isUnder)
+        (alf.ite (alf.is_eq n n1)
+          (alf.and (alf.not isUnder) ($str_arith_entail_simple n1))
+        (alf.ite (alf.is_eq n (+ n1 1))
+          (alf.and (alf.not isUnder) ($str_arith_entail_simple_pred (> n1 0)))
+        (alf.ite (alf.is_eq n 1)
+          (alf.and isUnder ($str_arith_entail_simple n1))
+          false))))
+  )
+)
+
+; program: $str_arith_entail_is_approx_len
+; args:
+; - s String: The first string argument to str.indexof.
+; - t String: The second string argument to str.indexof.
+; - n Int: The third argument of str.indexof.
+; - m Int: An integer that may under (resp. over) approximate the given indexof term when b is true (resp. false).
+; - b Bool: Whether we are checking that m is an under (resp. over) approximation.
+; return: >
+;   true if we can show that (>= (str.indexof s t n) m) (or <= if b is false) holds using
+;   approximation cases as defined by Figure 2 of "High-Level Abstractions for Simplifying
+;   Extended String Constraints in SMT", CAV 2019. When applicable, we also rely on calls
+;   that reason that string length is non-negative and basic properties of arithmetic.
+(define $str_arith_entail_is_approx_indexof ((s String) (t String) (n Int) (m Int) (isUnder Bool))
+  (let ((ls (str.len s)))
+  (let ((lt (str.len t)))
+  (alf.ite (alf.is_eq m -1)
+    isUnder
+  (alf.ite (alf.is_eq m ls)
+    (alf.not isUnder)
+  (alf.ite (alf.is_eq m (- ls lt))
+    (alf.and (alf.not isUnder) ($str_arith_entail_simple_pred (>= ls lt)))
+    false))))))
+
+; define: $str_arith_entail_is_approx_to_int
+; args:
+; - s String: The first string argument to str.to_int.
+; - n Int: An integer that may under (resp. over) approximate the given indexof term when b is true (resp. false).
+; - b Bool: Whether we are checking that n is an under (resp. over) approximation.
+; return: >
+;   true if we can show that (>= (str.to_int s) n) (or <= if b is false) holds using approximation cases
+;   as defined by Figure 2 of "High-Level Abstractions for Simplifying Extended String Constraints
+;   in SMT", CAV 2019.
+(define $str_arith_entail_is_approx_to_int ((s String) (n Int) (isUnder Bool))
+  (alf.ite (alf.is_eq n -1) isUnder false)
+)
+
+; define: $str_arith_entail_is_approx
+; args:
+; - n Int: The first integer term.
+; - m Int: The second integer term.
+; - b Bool: Whether we are checking that n is an under (resp. over) approximation of m.
+; return: >
+;   true if we can show that (>= n m) (resp (<= n m) when b is false) holds using approximation cases
+;   as defined by Figure 2 of "High-Level Abstractions for Simplifying Extended String Constraints
+;   in SMT", CAV 2019.
+(program $str_arith_entail_is_approx ((s String) (t String) (n1 Int) (n2 Int :list) (n3 Int) (n4 Int :list) (n5 Int) (isUnder Bool))
```

Comment:
I wish we could just flip the arguments to avoid isUnder 😩

Reply:
Again this is something that I'd prefer doing in the solver first before we modify the signature.